src/DependencyGraph/DependencyGraphHelpers.js

```diff
@@ -8,26 +8,115 @@
 
+const NODE_MODULES_DIR = path.sep + 'node_modules' + path.sep;
+
 class DependencyGraphHelpers {
-  constructor({ providesModuleNodeModules, assetExts }) {
-    this._providesModuleNodeModules = providesModuleNodeModules;
+  constructor({
+    roots = [],
+    providesModuleNodeModules = [],
+    assetExts = [],
+  }) {
+    this._hasteRegex = this._buildHasteRegex(
+      this._resolveHastePackages(providesModuleNodeModules, roots)
+    );
     this._assetExts = assetExts;
   }
 
+  /**
+   * An item in the `providesModuleNodeModule` array is an object containing two keys:
+   *  * 'name' - the package name
+   *  * 'parent' - if a string, it's the name of the parent package of the module in
+   *  the dep tree. If null, it signifies that we should look for this package at
+   *  the top level of the module tree.
+   *
+   * We try to resolve the specified module...but if we can't, we fallback to looking
+   * in 'node_modules/[package name]' (this satisfies use cases where the actual
+   * 'react-native' package is one of the root dirs, such as when we run tests
+   * or examples).
+   */
+  _resolveHastePackages(packages, roots) {
+    const packagePathForPackage = ({ name, parent }, rootDir) => {
+      let packagePath;
+      if (parent) {
+        if (!Array.isArray(parent)) {
+          parent = [parent];
+        }
+        parent.push(name);
+        packagePath = rootDir +
+          NODE_MODULES_DIR +
+          parent.join(NODE_MODULES_DIR);
+      } else {
+        packagePath = rootDir + NODE_MODULES_DIR + name;
+      }
+
+      if (packagePath.endsWith(path.sep)) {
+        return packagePath.slice(0, -1);
+      } else {
+        return packagePath;
+      }
+    };
+
+    return packages.reduce((hastePackages, p) => {
+      roots.forEach(rootDir => {
+        const packagePath = packagePathForPackage(p, rootDir);
+        try {
+          const stats = fs.statSync(packagePath);
+          if (stats && stats.isDirectory()) {
+            hastePackages.push(packagePath);
+          }
+        } catch (e) {
+          // if we don't find the package, let's just default to node_modules/[package name]
+          hastePackages.push(packagePathForPackage({ name: p.name }, rootDir));
+        }
+      });
+      return hastePackages;
+    }, []);
+  }
+
+  /**
+   * Given a list of directories, build a regex that takes the form:
+   * 	^((?![module's node_modules dir])[module dir])|
+   * 	 ((?![next module's node_modules dir])[next module dir])|...
+   *
+   * This is an alternative to looping through any of the providesModuleNodeModules
+   * during `isNodeModulesDir`, which is run tens of thousands of times in a typical
+   * project. A regex is much faster in this use-case.
+   */
+  _buildHasteRegex(dirs) {
+    const dirRegexes = dirs.reduce((regexes, dir) => {
+      const dirRegex = '((?!' +
+        escapeRegExp(dir + NODE_MODULES_DIR) + ')' +
+        escapeRegExp(dir + path.sep) + ')';
+      regexes.push(dirRegex);
```

Comment:
not sure you are getting much mileage out of this reduce, forEach + an outer `regexes` variable seems like one line less and added readability? :P